### DISCUSSION TITLE
How to parse database reader XML result (1..n) into one HL7 message?

### DISCUSSION THREAD

**Discussion by areutera** [nextgenhealthcare/connect#6182] [Q&A]
Following up my [previous question](https://github.com/nextgenhealthcare/connect/discussions/6161), I successfully wrote the fields I wanted into a database.

On retrieving them, they are appear in an XML format like this (this is kept very simple to figure out the entire process):

<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<result>
    <id>11</id>
    <messageid>77484805</messageid>
    <segment>PV1</segment>
    <element>19</element>
    <subelement>1</subelement>
    <content>24020188</content>
</result>

In the end I want to fill certain fields in an outgoing HL7 2.5 message with the content of several of these elements and I am a bit stumped on how to approach this. 

1. Every single line in my (currently very simple) SQL SELECT statement produces one message, which in turn tries to produce one HL7 message. My intention, though, is to combine several of these results into one single HL7 message.
2. Currently I don't have an idea on how to capture these results (e.g. in a channel map) to then build the message. The channel Source is a database reader, which is "busy" with getting the data. The Destinations tab theoretically would allow me to parse the content into a channel map over the course of several steps in any one Destination I set... I believe. But it is not all apparent, how that could work, especially with getting the data from several messages.

Does anybody have an idea to get me started on the right track? Right now it's confusing as heck,

**Comment by pacmano1** [comment]
Change your source SQL to return an array of results as a single message.

**Reply by areutera** [reply]
That makes sense. I can then loop through the array, but where does the Destinations tab give me the space to put such a JavaScript program and how do I access the array? Is there any documentation on that? The User manual only gave me a "list of buttons" on the UI that I see, and not much on how to use it...

If there are threads here that deal with this, I'd be happy to have them pointed out to me as well. Right now, though, I don't even know how to properly phrase any search query. :(

**Reply by pacmano1** [reply]
firstly, you really need to take training.

Both source and destination transformers are designed for this.  You can also use filters to modify messages.

**Reply by pacmano1** [reply]
Please go over to the mirth forums site (not this one).  Over there is years of history of q and a

**Reply by areutera** [reply]
Thanks for your quick answers (national holiday here yesterday)!

While I totally understand where you are coming from (and agree with you), training in Mirth is hard to come by over here, especially within the time frame I am looking at.

I am aware of the forum I mentioned, but as I pointed out, I am trying to get a a sense of what key words might actually enable me to _find_ something worthwhile. Up to now I have not been successful doing that, hence me asking here.

Once again, thank you.

**Comment by tonygermano** [comment]
https://docs.nextgen.com/bundle/Mirth_User_Guide_4_3_0/page/connect/connect/topics/c_Database_Reader_connect_ug.html

The Database Reader has an option called "aggregate results" which will return the entire resultset as a single message instead of a message per row.

**Reply by areutera** [reply]
Yes, thank you, I did in fact find the aggregate results option!

It returns a result like this

<results>
    <result>
        <id>6</id>
        <messageid>77484805</messageid>
        <segment>MSH</segment>
        <element>7</element>
        <subelement>1</subelement>
        <content>20240214091932</content>
    </result>
    <result>
...
    </result>
...
</results>

which I can see in the Raw Source of each query.

My current problem is (and I have been looking for this quite a bit), that I want to access this XML object (if it is saved as such) in the Source Transformer, but I am having trouble figuring out, how to access it (for a HL7 message this would be msg[...][...], but what does Mirth call this custom object?

The outbound object I can see in the log file using msg.toString().

**Reply by tonygermano** [reply]
If your db query is returning a hl7 message as a string, you'll need to use the SerializerFactory to manually convert the hl7 message to xml. There should be plenty of examples if you search in github or the old forums.

**Reply by tonygermano** [reply]
SerializerFactory is part of the User API

https://docs.nextgen.com/bundle/Mirth_User_Guide_411/page/connect/connect/topics/c_The_User_API_Javadoc_connect_ug.html
